walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,102 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+
+    // first iterate through and combine events that overlap
+    // subtract combined events calendar from whole day
+    // remove time ranges that are too short 
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+        List<TimeRange> avaliableRanges = new ArrayList<TimeRange>();
+        avaliableRanges.add(TimeRange.WHOLE_DAY);
+        Collection<TimeRange> bookedRanges = combinedMandatoryRanges(events, request);
+
+        for (TimeRange booked : bookedRanges) {
+            TimeRange lastInAvaliable = avaliableRanges.get(avaliableRanges.size()-1);
+            if (lastInAvaliable.overlaps(booked)){
+                List<TimeRange> newLastAvaliable = removeOverlap(lastInAvaliable, booked);
+                avaliableRanges.remove(lastInAvaliable);
+                avaliableRanges.addAll(newLastAvaliable);
+                Collections.sort(avaliableRanges, TimeRange.ORDER_BY_START);
```

Comment:
This may result in a higher time complexity than you want. Consider using a subclass of NavigableSet (since available ranges are all disjoint) to keep time ranges sorted (https://docs.oracle.com/javase/7/docs/api/java/util/NavigableSet.html)

Reply:
I actually don't need to call sort here, because it should be already sorted and I put this line here as a precaution. I'll just remove this line.